### PR TITLE
KEP 3589: Uniformly filter manageJobsWithoutQueueNames by namespace

### PR DESCRIPTION
**Pull request by dgrove-oss**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

KEP to propose a design for #3589 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: keps/3589-manage-jobs-selectively/README.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+All unit tests for the current Pod integration namespaceSelector will be extended across all integrations.
+
+#### Integration tests
+
+All integration tests for the current Pod integration namespaceSelector will be extended across all integrations.
+
+### Graduation Criteria
+
+The feature is implemented as described for all Job integrations.
```
Comment (kannon92) [inline keps/3589-manage-jobs-selectively/README.md:128]: I don't think this really answers the question for graduation critiera.

This would usually be what would be required to "graduate" this feature to GA.

I'm not sure if this needs a beta stage..

File: keps/3589-manage-jobs-selectively/README.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+`namespaceSelector` is used by the Pod integration.
+
+We migrate away from using `podOptions.namespaceSelector` over the course of several releases.
+Ideally we remove it from the `Configuration` struct as part of going to the `v1` API level.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Cluster admins will be able to use `manageAllJobsNamespaceSelector` to easily configure Kueue
```
Comment (kannon92) [inline keps/3589-manage-jobs-selectively/README.md:73]: Reading this, I don't quite understand the interaction of ManageJobsWithoutQueueName and ManageAllJobsNamespaceSelector. Could you give an example of how you would configure this and how it would work?

I'm just not entirely sure what someone would put for a label selector in this case.

File: keps/3589-manage-jobs-selectively/README.md (hunk 1 above)
Comment (dgrove-oss) [reply keps/3589-manage-jobs-selectively/README.md]: I dropped this section.

File: keps/3589-manage-jobs-selectively/README.md (hunk 2 above)
Comment (dgrove-oss) [reply keps/3589-manage-jobs-selectively/README.md]: I expanded the proposal and user story to address this feedback.

File: keps/3589-manage-jobs-selectively/README.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+use of ValidatingAdmissionPolicies to ensure that all Kinds that could be managed by Kueue that
+are created in "user" namespaces have a queue name label (and thus will be managed by Kueue).
+
+However, while exploring this idea further a complication arose. The ValidatingAdmissionPolicy
+must be able to distinguish between top-level resources without queue names (which should be denied)
+and child resources that are owned by Kueue-managed parents (which must be admitted even if they do not have a queue name).
+In other words, the VAP must implement the same functionality as `jobframework.IsOwnerManagedByKueue`
+and cross check controller-references against managed kinds.  This approach may be challenging to scale
+as the number of Kueue managed kinds increases and ancestor trees become deeper.
+For example in the not so distant future we might see: AppWrapper->PyTorchJob->JobSet->Job->Pod.
```
Comment (KPostOffice) [inline keps/3589-manage-jobs-selectively/README.md:189]: Wouldn't it be enough to do the same as is done in `jobframework.IsOwnerManagedByKueue` and just check if the parent is manageable by Kueue? So ancestor tree size wouldn't matter.

File: keps/3589-manage-jobs-selectively/README.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+```
+
+### Implementation
+
+The configuration of all Job webhooks are extended to propagate the namespace selector and it is
+used to filter `manageJobsWithoutQueueName` as is currently done in the Pod integration. Specifically,
+jobs without queue names are only managed if both `manageJobsWithoutQueueName` is true and
+the jobs namespace matches the `managedJobsNamespaceSelector`.
+
+Configuration parsing during startup is extended to give a deprecation notice if `podOptions.namespaceSelector` is set.
```
Comment (KPostOffice) [inline keps/3589-manage-jobs-selectively/README.md:141]: One note, the proposed configuration doesn't have a way of impersonating `podOptions.podSelector`. Would we be okay with deprecating it anyways?

File: keps/3589-manage-jobs-selectively/README.md (hunk 3 above)
Comment (dgrove-oss) [reply keps/3589-manage-jobs-selectively/README.md:189]: I will try to phrase this better.  It isn't that the check has to actually crawl up the ancestor tree.  It's that the check has to consider each of the kinds that could be Kueue-managed parents.  So it needs a longer list of kinds and it has to somehow know that a particular Kind's Kueue integration is enabled.   For kinds like Job or JobSet that could have multiple Kueue-managed parent Kinds, each has to be covered by the check.

File: keps/3589-manage-jobs-selectively/README.md (hunk 4 above)
Comment (dgrove-oss) [reply keps/3589-manage-jobs-selectively/README.md:141]: I think `podOptions.podSelector` is orthogonal to `podOptions.namespaceSelector`.  This KEP is only proposing to deprecate `podOptions.namespaceSelector` in favor of the new namespaceSelector that would apply to all integrations.

File: keps/3589-manage-jobs-selectively/README.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+	// ManageJobsWithoutQueueName controls whether or not Kueue reconciles
+	// jobs that don't set the label kueue.x-k8s.io/queue-name.
+	// If set to true, then those jobs will be suspended and never started unless
+	// they are assigned a queue and eventually admitted. This also applies to
+	// jobs created before starting the kueue controller.
+	// Defaults to false; therefore, those jobs are not managed and if they are created
+	// unsuspended, they will start immediately.
+	ManageJobsWithoutQueueName bool `json:"manageJobsWithoutQueueName"`
+
+  // ManageAllJobsNamespaceSelector can be used to omit some namespaces from ManageJobsWithoutQueueName
```
Comment (mimowo) [inline keps/3589-manage-jobs-selectively/README.md:127]:
```suggestion
  // ManagedJobsNamespaceSelector can be used to omit some namespaces from ManagedJobsWithoutQueueName
```
nit, to align the naming

File: keps/3589-manage-jobs-selectively/README.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+### Validating Admission Policies
+
+As previously discussed in [issue 2119](https://github.com/kubernetes-sigs/kueue/issues/2119),
+we could phase out `manageJobsWithoutQueueNames` and instead recommend the
+use of ValidatingAdmissionPolicies to ensure that all Kinds that could be managed by Kueue that
+are created in "user" namespaces have a queue name label (and thus will be managed by Kueue).
+
+However, while exploring this idea further a complication arose. The ValidatingAdmissionPolicy
+must be able to distinguish between top-level resources without queue names (which should be denied)
+and child resources that are owned by Kueue-managed parents (which must be admitted even if they do not have a queue name).
```
Comment (mimowo) [inline keps/3589-manage-jobs-selectively/README.md:185]: Yes, I can understand why VAP becomes complicated when there are long chains of CRDs. I think this picture may be even more complex than in this example if we have multiple parent CRDs which can own batch/v1 Job.

Also, introducing `ManagedJobsNamespaceSelector` isn't really adding new complexity assuming we deprecate and drop podOptions.namespaceSelector over time (another comment)

**Comment by mimowo** [comment]
LGTM overall, I would just like to use this work to deprecate the podOptions.namespaceSelector, so we don't need to maintain both APIs in the long run. Then, I think the API is just a more generic replacement to what we currently have, but I don't consider it much of a complication as we already do it for pods.

**Comment by tenzen-y** [comment]
Landing on here

/cc

File: keps/3589-manage-jobs-selectively/README.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+Ideally we remove it from the `Configuration` struct as part of going to the `v1` API level.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Cluster admins will be able to use the combination of `managedJobsNamespaceSelector` and
+`manageJobsWithoutQueueName` to configure Kueue to enable quota enforcement on a per namespace basis.
+
+One sapproach would be to exempt selected namespaces from management:
```
Comment (tenzen-y) [inline keps/3589-manage-jobs-selectively/README.md:93]:
```suggestion
One approach would be to exempt selected namespaces from management:
```

File: keps/3589-manage-jobs-selectively/README.md [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+managedJobsNamespaceSelector:
+  matchLabels:
+    kueue.x-k8s.io/managed-namespace: true
+```
+
+## Design Details
+
+### API
+
+The `Configuration` struct is extended to add ManageAllJobsNamespaceSelector
```
Comment (tenzen-y) [inline keps/3589-manage-jobs-selectively/README.md:112]:
```suggestion
The `Configuration` struct is extended to add ManagedJobsNamespaceSelector
```

File: keps/3589-manage-jobs-selectively/README.md [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+would expect them to be configured in the same way.
+
+### Goals
+
+Provide a simple and uniform API to enable cluster admins to configure Kueue
+such that quotas will be enforced for specific namespaces.
+
+### Non-Goals
+
+Consider any scope for filtering `manageJobsWithoutQueueName` that is not based on namespaces.
```
Comment (tenzen-y) [inline keps/3589-manage-jobs-selectively/README.md:68]: This is just comment.

My primary concern for this KEP was this one.
I would not like to accept this feature in any case since this tries to re-implement the Gatekeeper, VAP, and MAP.

File: keps/3589-manage-jobs-selectively/README.md [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+
+### API
+
+The `Configuration` struct is extended to add ManageAllJobsNamespaceSelector
+```go
+type Configuration struct {
+
+  ...
+
+	// ManageJobsWithoutQueueName controls whether or not Kueue reconciles
```
Comment (tenzen-y) [inline keps/3589-manage-jobs-selectively/README.md:118]: This is just a comment and not an objection to this feature.

I was wondering if we could just deprecate and remove the `manageJobsWithoutQueueName` and introduce more informable validations so that batch users can easily find the quota violation.

I think that the current manageJobsWithoutQueuenName does not have a good user experience since there is no way to find the reason why jobs are not runnable when they accidentally submit Jobs without queue names.

Because the Kueue automatically suspends the Jobs.

However, based on some consideration, this `manageJobsWithoutQueueName` has already been released. So, even if we do not implement this, we must follow the appropriate deprecation steps.

Additionally, if the cluster administrator provides the automatic queue-name insertion mechanism, this could be helpful for the batch users.
Furthermore, such an informable validation mechanism could be an alternative new feature since validation and mutation could not cover all use cases each other, and I think that both are valuable.

File: keps/3589-manage-jobs-selectively/README.md [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+In particular, there must be a robust and recommended mechanism for configuring Kueue
+such that all Pod-creating Kinds in user namespaces are subject to management by Kueue.
+
+Currently, the primary mechanism for doing this is setting `manageJobsWithoutQueueName` to
+true.  When `manageJobsWithoutQueueName` is true, then Kueue will manage all instances of
+Kinds with Kueue integrations, even if they do not have a `kueue.x-k8s.io/queue-name` label.
+As a practical matter, Pods running in system namespaces must be exempted from being
+managed by Kueue even if `manageJobsWithoutQueueName` is true.  Therefore Kueue's Pod
+integration limits the scope of `manageJobsWithoutQueueName` by restricting its effect to
+Pods that are in namespaces that match against the `integrations.podOptions.namespaceSelector`.
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply keps/3589-manage-jobs-selectively/README.md:45]: I think that we can just mark podNamespaceSelector as deprecated.

[the remark this answers is not in the document]
Comment (mimowo) [reply keps/3589-manage-jobs-selectively/README.md:45]: If we move forward with this KEP we can. Otherwise we still need it to say that pod integration, deployement, and statefulSet is constrained.

File: keps/3589-manage-jobs-selectively/README.md (hunk 6 above)
Comment (dgrove-oss) [reply keps/3589-manage-jobs-selectively/README.md:185]: I think this is the key point.  For either `manageJobsWithoutQueueName` or the default LocalQueue name being pursued in #3610 to be usable, there _has_ to be a way to control which namespaces they apply to for at least the `Pod`, `StatefulSet` and `Deployment` integrations.   On OpenShift, which tends to use `Job` for system operations the namespace control needs to also cover the `Job` integration as well. 

I can rework this KEP to focus on just the generalization of `integrations.podOptions.namespaceSelector` to `managedJobsNamespaceSelector`.    I think we need it independent of what happens to `manageJobsWithoutQueueName`.

File: keps/3589-manage-jobs-selectively/README.md (hunk 9 above)
Comment (mimowo) [reply keps/3589-manage-jobs-selectively/README.md:68]: Yes, I fully agree we don't want this mechanism to be more complex, and +1 to reject features not based on namespaces, but for namespaces we already have the complication in `podOptions.namespaceSelector`, and making it uniformly supported seems like good idea when we have Deployments and StatefulSets supported.

I agree with the direction to use VAP and MAP, but this will probably take many releases as MAP is still alpha in 1.32, and we need to propose details for VAP and MAP generation to avoid issues with hierarchical jobs (as described in the KEP).

Comment (tenzen-y) [reply keps/3589-manage-jobs-selectively/README.md:68]: Thanks for expanding that. If anyone wants to know VAP and MAP introduction details, please refer https://github.com/kubernetes-sigs/kueue/issues/3636 and https://github.com/kubernetes-sigs/kueue/issues/3637.

File: keps/3589-manage-jobs-selectively/README.md [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+We use this new configuration for all integrations in the same way that the current
+`namespaceSelector` is used by the Pod integration.  Specifically,
+1. If `manageJobsWithoutQueueName` is false, `managedJobsNamespaceSelector` has no effect: Kueue will manage
+exactly those instances of supported Kinds that have a `queue-name` label.
+2. If `manageJobsWithoutQueueName` is true, then Kueue will (a) manage all instances of supported Kinds
+that have a `queue-name` label and (b) will manage all instances of supported Kinds that do not
+have a `queue-name` label if they are in namespaces that match `managedJobsNamespaceSelector`.
+
+We migrate away from using `podOptions.namespaceSelector` over the course of several releases.
+Ideally we remove it from the `Configuration` struct as part of going to the `v1` API level.
```
Comment (mimowo) [inline keps/3589-manage-jobs-selectively/README.md:84]: Sgtm, I think we can say that in the transition period the actual namespace selector used for Pod integration (and thus StatefulSet and Deployments) will be the intersection of both `managedJobsNamespaceSelector` and `podOptions.namespaceSelector`.

**Review by mimowo — COMMENTED** [review]
LGTM overall, left a remark to explain how the effective namespace selector in the transition period before we drop podOptions.namespaceSelector

**Comment by dgrove-oss** [comment]
Now that there is actually a KEP for 2936, I've updated this KEP accordingly (in particular drawbacks and alternatives section). 

The TL;DR is that if we pursue 2936 and it will implicitly filter by namespace (which seems likely since it is looking for a default local queue in the job's namespace) then we don't need to pursue this KEP.

We may still want to pursue this KEP as a tactical move, if we think it will take a significant number of releases to get 2936 to a usable state and really remove `manageJobsWithoutQueueName`.  I think the implementation effort for this KEP is fairly low (and as discussed in alternatives could also be done without an API change if we are ok with an even odder semantics for `podOptions.namespaceSelector` until we remove it entirely).

/cc @mimowo @tenzen-y @kannon92 @yaroslava-serdiuk

**Comment by mimowo** [comment]
I think this KEP is still useful as a generalization of `podOptions.namespaceSelector`, which is needed as long as we support `manageJobsWithoutQueueName`.

Maybe one day with LQ defaulting and generated VAP we could drop `manageJobsWithoutQueueName: true`, but this seems a long way off. So, I think this PR makes sense as a tactical move.

@dgrove-oss do you think if we accept the KEP you could provide implementation for 0.10? This is planned 3-4  Dec.

**Comment by tenzen-y** [comment]
> I think this KEP is still useful as a generalization of podOptions.namespaceSelector, which is needed as long as we support manageJobsWithoutQueueName.
> 
> Maybe one day with LQ defaulting and generated VAP we could drop manageJobsWithoutQueueName: true, but this seems a long way off. So, I think this PR makes sense as a tactical move.

I have the same opinion with Michal. Regardless of #2936, this sounds useful.
In addition to Michal mentioned use cases, I can images the useful situations where the cluster admin want to use the dedicated LQ name separate from "default", and cluster admin want to obviously tell batch users that all Jobs should be obviously enqueue to specific LocalQueue.

**Comment by dgrove-oss** [comment]
I have time on Monday/Tuesday to work on an implementation.  If it goes smoothly, that may be enough.  I can give it a try.

**Comment by mimowo** [comment]
/lgtm
Leaving approval tag to @tenzen-y

**Review by tenzen-y — COMMENTED** [review]
Thank you!
/approve
